File: FirstApp/ServerManagement/Form7.vb
Imports System.Text.RegularExpressions
Imports Renci.SshNet
Imports Renci.SshNet.Common
Imports System.Runtime.CompilerServices


Public Class Form7
    Private Sub Form7_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Dim client = New Ssh()
        Dim output = client.Execute("ls")
    End Sub
End Class

Public Class Log
    Public Shared Sub Verbose(message As String)
        Console.WriteLine(message)
    End Sub

    Public Shared Sub [Error](message As String)
        Console.WriteLine(message)
    End Sub
End Class

Module StringExt
    <Extension()>
    Function StringBeforeLastRegEx(str As String, regex As Regex) As String
        Dim matches = regex.Matches(str)
        Return If(matches.Count > 0, str.Substring(0, matches.Item(matches.Count - 1).Index), str)
    End Function

    <Extension()>
    Function EndsWithRegEx(str As String, regex As Regex) As Boolean
        Dim matches = regex.Matches(str)
        Return matches.Count > 0 AndAlso str.Length = (matches.Count - 1 + matches.Item(matches.Count - 1).Length)
    End Function

    <Extension()>
    Function StringAfter(str As String, substring As String) As String
        Dim index = str.IndexOf(substring, StringComparison.Ordinal)
        Return If(index >= 0, str.Substring(index + substring.Length), "")
    End Function

    <Extension()>
    Function GetLines(str As String) As String()
        Return Regex.Split(str, vbCrLf & "|" & vbCr & "|" & vbLf)
    End Function
End Module

Module UtilExt
    <Extension()>
    Sub ForEach(Of T)(sequence As IEnumerable(Of T), func As Action(Of T))
        For Each item In sequence
            func(item)
        Next
    End Sub
End Module

Public Class Ssh
    Private sshClient As SshClient
    Private shell As ShellStream
    Private pwd As String = "tapus123"
    Private lastCommand As String = ""
    Shared prompt As Regex = New Regex("[a-zA-Z0-9_.-]*\@[a-zA-Z0-9_.-]*\:\~[#$] ", RegexOptions.Compiled)
    Shared pwdPrompt As Regex = New Regex("password for .*\:", RegexOptions.Compiled)
    Shared promptOrPwd As Regex = New Regex(Ssh.prompt.ToString + "|" + Ssh.pwdPrompt.ToString)

    Public Sub Connect()
        Dim host = "3.16.198.57"
        Dim username = "ec2-user"
        Dim password = "tapus123"

        sshClient = New SshClient(host, username, password)
        sshClient.Connect()
        Dim terminalMode = New Dictionary(Of TerminalModes, UInteger)()
        terminalMode.Add(TerminalModes.ECHO, 53)
        shell = sshClient.CreateShellStream("", 0, 0, 0, 0, 4096, terminalMode)

        Try
            Expect(Ssh.prompt)
        Catch ex As Exception
            Log.[Error]("Exception - " & ex.Message)
            Throw
        End Try
    End Sub

    Public Sub Disconnect()
        Log.Verbose($"Ssh Disconnect")
        sshClient?.Disconnect()
        sshClient = Nothing
    End Sub

    Private Sub Write(commandLine As String)
        Console.ForegroundColor = ConsoleColor.Green
        Log.Verbose("> " & commandLine)
        Console.ResetColor()
        lastCommand = commandLine
        shell.WriteLine(commandLine)
    End Sub

    Private Function Expect(_expect As Regex, Optional timeoutSeconds As Double = 60.0) As String
        Dim result = shell.Expect(_expect, TimeSpan.FromSeconds(timeoutSeconds))

        If result Is Nothing Then
            Throw New Exception($"Timeout {timeoutSeconds}s executing {lastCommand}")
        End If

        result = If(result.Contains(lastCommand), result.StringAfter(lastCommand), result)
        result = result.StringBeforeLastRegEx(Ssh.prompt)
        result = result.Trim()
        result.GetLines().ForEach(Sub(x) Log.Verbose(x))
        Return result
    End Function

    Public Function Execute(commandLine As String, Optional timeoutSeconds As Double = 30.0) As String
        Dim exception As Exception = Nothing
        Dim result = ""
        Dim errorMessage = "failed"
        Dim errorCode = "exception"

        Try
            Write(commandLine)
            result = Expect(Ssh.promptOrPwd)

            If result.EndsWithRegEx(pwdPrompt) Then
                Write(pwd)
                Expect(Ssh.prompt)
            End If

            Write("echo $?")
            errorCode = Expect(Ssh.prompt)

            If errorCode = "0" Then
                Return result
            ElseIf result.Length > 0 Then
                errorMessage = result
            End If

        Catch ex As Exception
            exception = ex
            errorMessage = ex.Message
        End Try

        Throw New Exception($"Ssh error: {errorMessage}, code: {errorCode}, command: {commandLine}", exception)
    End Function
End Class

File: OrnaTrack/ServerManagement/Form7.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Microsoft.VisualBasic;
using Renci.SshNet;
using Renci.SshNet.Common;

namespace FirstApp
{


    public partial class Form7
    {
        public Form7()
        {
            InitializeComponent();
        }
        private void Form7_Load(object sender, EventArgs e)
        {
            var client = new Ssh();
            string output = client.Execute("ls");
        }
    }

    public class Log
    {
        public static void Verbose(string message)
        {
            Console.WriteLine(message);
        }

        public static void Error(string message)
        {
            Console.WriteLine(message);
        }
    }

    static class StringExt
    {
        public static string StringBeforeLastRegEx(this string str, Regex regex)
        {
            var matches = regex.Matches(str);
            return matches.Count > 0 ? str.Substring(0, matches[matches.Count - 1].Index) : str;
        }

        public static bool EndsWithRegEx(this string str, Regex regex)
        {
            var matches = regex.Matches(str);
            return matches.Count > 0 && str.Length == matches.Count - 1 + matches[matches.Count - 1].Length;
        }

        public static string StringAfter(this string str, string substring)
        {
            int index = str.IndexOf(substring, StringComparison.Ordinal);
            return index >= 0 ? str.Substring(index + substring.Length) : "";
        }

        public static string[] GetLines(this string str)
        {
            return Regex.Split(str, Constants.vbCrLf + "|" + Constants.vbCr + "|" + Constants.vbLf);
        }
    }

    static class UtilExt
    {
        public static void ForEach<T>(this IEnumerable<T> sequence, Action<T> func)
        {
            foreach (var item in sequence)
                func(item);
        }
    }

    public class Ssh
    {
        private SshClient sshClient;
        private ShellStream shell;
        private string pwd = "tapus123";
        private string lastCommand = "";
        private static Regex prompt = new Regex(@"[a-zA-Z0-9_.-]*\@[a-zA-Z0-9_.-]*\:\~[#$] ", RegexOptions.Compiled);
        private static Regex pwdPrompt = new Regex(@"password for .*\:", RegexOptions.Compiled);
        private static Regex promptOrPwd = new Regex(prompt.ToString() + "|" + pwdPrompt.ToString());

        public void Connect()
        {
            string host = "3.16.198.57";
            string username = "ec2-user";
            string password = "tapus123";

            sshClient = new SshClient(host, username, password);
            sshClient.Connect();
            var terminalMode = new Dictionary<TerminalModes, uint>();
            terminalMode.Add(TerminalModes.ECHO, 53U);
            shell = sshClient.CreateShellStream("", 0U, 0U, 0U, 0U, 4096, terminalMode);

            try
            {
                Expect(prompt);
            }
            catch (Exception ex)
            {
                Log.Error("Exception - " + ex.Message);
                throw;
            }
        }

        public void Disconnect()
        {
            Log.Verbose($"Ssh Disconnect");
            sshClient?.Disconnect();
            sshClient = null;
        }

        private void Write(string commandLine)
        {
            Console.ForegroundColor = ConsoleColor.Green;
            Log.Verbose("> " + commandLine);
            Console.ResetColor();
            lastCommand = commandLine;
            shell.WriteLine(commandLine);
        }

        private string Expect(Regex _expect, double timeoutSeconds = 60.0d)
        {
            string result = shell.Expect(_expect, TimeSpan.FromSeconds(timeoutSeconds));

            if (result is null)
            {
                throw new Exception($"Timeout {timeoutSeconds}s executing {lastCommand}");
            }

            result = result.Contains(lastCommand) ? result.StringAfter(lastCommand) : result;
            result = result.StringBeforeLastRegEx(prompt);
            result = result.Trim();
            result.GetLines().ForEach(x => Log.Verbose(x));
            return result;
        }

        public string Execute(string commandLine, double timeoutSeconds = 30.0d)
        {
            Exception exception = null;
            string result = "";
            string errorMessage = "failed";
            string errorCode = "exception";

            try
            {
                Write(commandLine);
                result = Expect(promptOrPwd);

                if (result.EndsWithRegEx(pwdPrompt))
                {
                    Write(pwd);
                    Expect(prompt);
                }

                Write("echo $?");
                errorCode = Expect(prompt);

                if (errorCode == "0")
                {
                    return result;
                }
                else if (result.Length > 0)
                {
                    errorMessage = result;
                }
            }

            catch (Exception ex)
            {
                exception = ex;
                errorMessage = ex.Message;
            }

            throw new Exception($"Ssh error: {errorMessage}, code: {errorCode}, command: {commandLine}", exception);
        }
    }
}